smokey/src/main/java/frc/robot/common/CorrectableEncoderRevNeoPlusDigitalIoPort.java

```diff
@@ -0,0 +1,80 @@
+// ************************************************************
+// Bishop Blanchet Robotics
+// Home of the Cybears
+// FRC - Charged Up - 2023
+// File: public enum CorrectableEncoderRevNeoPlusDigitalIoPort.java
+// Intent: Forms an interface that will help keep a motor encoder from drifting with use of external sensor.
+// ************************************************************
+
+// ʕ •ᴥ•ʔ ʕ•ᴥ•  ʔ ʕ  •ᴥ•ʔ ʕ •`ᴥ´•ʔ ʕ° •° ʔ ʕ •ᴥ•ʔ ʕ•ᴥ•  ʔ ʕ  •ᴥ•ʔ ʕ •`ᴥ´•ʔ ʕ° •° ʔ 
+
+package frc.robot.common;
+
+import com.revrobotics.RelativeEncoder;
+
+import edu.wpi.first.wpilibj.DigitalInput;
+
+public class CorrectableEncoderRevNeoPlusDigitalIoPort implements ICorrectableEncoder {
+
+    private RelativeEncoder revNeoMotorEncoder = null;
+    private DigitalInput dioStateDevice = null;
+    private double encoderTicksAtStateChangeSetPoint = 0.0;
+    private boolean lastState = false;
+    private boolean motorEncoderPositionReset = false;
+
+    /**
+     * Concrete implementation of a pairing between motor encoder and DIO to coordinate resetting of 
+     * encoder position based on state change of DIO
+     * @param revNeoEncoder - the REV Neo motor encoder
+     * @param stateDevice - the DIO state device
+     * @param encoderTicksAtStateChange - the set point to use for the motors encoder when the state 
+     * @param initalizeEncoderPositionOnDioTriggered - when true and the device is currently triggered the initial position will be set 
+     * changes between the current and previous calls into getCurrentEncoderPosition
+     */
+    public CorrectableEncoderRevNeoPlusDigitalIoPort(
+        RelativeEncoder revNeoEncoder,
+        DigitalInput stateDevice,
+        double encoderTicksAtStateChange,
+        boolean initalizeEncoderPositionOnDioTriggered) {
+
+        revNeoMotorEncoder = revNeoEncoder;
+        dioStateDevice = stateDevice;
+        encoderTicksAtStateChangeSetPoint = encoderTicksAtStateChange;
+
+        lastState = dioStateDevice.get();
+
+        if(initalizeEncoderPositionOnDioTriggered == true && lastState == false) {
+            this.revNeoMotorEncoder.setPosition(this.encoderTicksAtStateChangeSetPoint);
+            motorEncoderPositionReset = true;
+        }
+    }
+
+    /**
+     * An method intended to be called periodically (as in 1x per 20 ms) that will be used to both
+     * correct the motors encoder (when necessary) and also return the current motor encoder position.
+     * @return a double representing the motor encoder ticks
+     */
+    public double getCurrentEncoderPosition() {
+
+        boolean currentState = this.dioStateDevice.get();
+
+        if(currentState != this.lastState) {
```

Comment:
One concern is that you reset the encoder position when the sensor changes state from false->true as well as true->false.  There will be some systematic error injected depending on which direction the arm traveled.  Would like to measure the positional variability in each of the 5 arm locations to ensure this gives us the precision we need.

Reply:
This is a good point, the implementation here assumes the sensor generally is triggered or not triggered at the same point of arm movement ... we can change it to be single direction if we want.

Reply:
The whole implementation makes the assumption that dual direction results in the same outcome.  If that is not true for this sensor, then we should likely go uni-directional (say some convention, like only on the downward direction).

Reply:
In the end, we really do need to test this and decide if we want different behavior.

Reply:
I think the current implementation will lead to repeatable positions.  Each position has the bottom of the tube either above or below the sensor.  For any position where the bottom of the tube is above the sensor, the position will be reset as the sensor travels upward.  And that position will always be consistent w.r.t. the position set while traveling upward. Vice-versa for positions where the bottom of the tube is below the sensor.  These positions will always be consistent w.r.t. the position set while traveling downward.  Let's confirm with testing.